File: 6. Zigzag Conversion/6. Zigzag Conversion.dart
class Solution {
  String convert(String s, int numRows) {
    if (numRows == 1) return s;
    
    List<StringBuffer> rows = List.generate(
      numRows.clamp(0, s.length),
      (_) => StringBuffer(),
    );
    
    int curRow = 0;
    bool goingDown = false;
    
    for (var c in s.runes) {
      rows[curRow].writeCharCode(c);
      if (curRow == 0 || curRow == numRows - 1) goingDown = !goingDown;
      curRow += goingDown ? 1 : -1;
    }
    
    return rows.map((sb) => sb.toString()).join();
  }
}

File: 6. Zigzag Conversion/6. Zigzag Conversion.rs
impl Solution {
    pub fn convert(s: String, num_rows: i32) -> String {
        let num_rows = num_rows as usize;
        if num_rows == 1 || num_rows >= s.len() {
            return s;
        }

        let mut rows: Vec<String> = vec![String::new(); num_rows];
        let mut current_row = 0;
        let mut step = 1;

        for c in s.chars() {
            rows[current_row].push(c);
            
            if current_row == 0 {
                step = 1;
            } else if current_row == num_rows - 1 {
                step = -1;
            }
            
            current_row = (current_row as i32 + step) as usize;
        }

        rows.concat()
    }
}